src/memory_region/mod.rs

```diff
@@ -0,0 +1,181 @@
+    ///
+    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
+    unsafe fn get_unchecked(self, slice: *const T) -> *const Self::Output;
+
+    /// Returns a mutable reference to the output at this location, without
+    /// performing any bounds checking.
+    /// Calling this method with an out-of-bounds index or a dangling `slice` pointer
+    /// is *[undefined behavior]* even if the resulting reference is not used.
+    ///
+    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
+    unsafe fn get_unchecked_mut(self, slice: *mut T) -> *mut Self::Output;
+
+    /// Returns a shared reference to the output at this location, panicking
+    /// if out of bounds.
+    fn index(self, slice: &T) -> &Self::Output;
+
+    /// Returns a mutable reference to the output at this location, panicking
+    /// if out of bounds.
+    fn index_mut(self, slice: &mut T) -> &mut Self::Output;
+}
+
+/// Mr slice Ptr Repr
+#[repr(C)]
+union MrSlicePtrRepr<T: ?Sized> {
+    /// const ptr
+    const_ptr: *const T,
+    /// mut ptr
+    mut_ptr: *mut T,
+    /// ptr and metadata components
+    components: MrSlicePtrComponents,
+}
+
+/// Mr slice components include Mr ptr and slice metadata
+#[repr(C)]
+#[derive(Clone, Copy)]
+struct MrSlicePtrComponents {
+    /// ptr to Mr
+    data_address: *const (),
+    /// metadata of mr slice
+    metadata: SliceMetaData,
+}
+
+/// get mr slice ptr from addr and memtadata
+#[inline]
+const fn mr_slice_from_raw_parts<T: ?Sized>(
+    data_address: *const (),
+    metadata: SliceMetaData,
+) -> *const T {
+    // SAFETY: Accessing the value from the `PtrRepr` union is safe since *const T
+    // and PtrComponents<T> have the same memory layouts. Only std can make this
+    // guarantee.
+    unsafe {
+        MrSlicePtrRepr {
+            components: MrSlicePtrComponents {
+                data_address,
+                metadata,
+            },
+        }
+        .const_ptr
+    }
+}
+
+/// get mut mr slice ptr from mut addr and memtadata
+#[inline]
+const fn mr_slice_from_raw_parts_mut<T: ?Sized>(
+    data_address: *mut (),
+    metadata: SliceMetaData,
+) -> *mut T {
+    // SAFETY: Accessing the value from the `PtrRepr` union is safe since *const T
+    // and PtrComponents<T> have the same memory layouts. Only std can make this
+    // guarantee.
+    unsafe {
+        MrSlicePtrRepr {
+            components: MrSlicePtrComponents {
+                data_address,
+                metadata,
+            },
+        }
+        .mut_ptr
+    }
+}
+
+/// get the metadata of mr slice
+#[inline]
+const fn mr_slice_metadata<T: ?Sized>(ptr: *const T) -> SliceMetaData {
+    // SAFETY: Accessing the value from the `PtrRepr` union is safe since *const T
+    // and PtrComponents<T> have the same memory layouts. Only std can make this
+    // guarantee.
+    unsafe { MrSlicePtrRepr { const_ptr: ptr }.components.metadata }
```

Comment:
The `MrSlicePtrRepr` is newly created on the stack, what's the metadata? I don't think this function works.